### PR TITLE
net/tlsdial: call out firewalls blocking Tailscale in health warnings

### PR DESCRIPTION
**Pull request by agottardo**
Updates tailscale/tailscale#13839

### CODE REVIEW

File: net/tlsdial/blockblame/blockblame.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+// Indicator represents a specific piece of information within a SSL certificate that can be
+// used to identify the manufacturer of a firewall.
+type Indicator struct {
+	IssuerName string // never empty, matches if Issuer contains this
+	EmailAddr  string // may be empty, matches if EmailAddresses contains this
+}
+
+var Manufacturers = []Manufacturer{
```
Comment (bradfitz) [inline net/tlsdial/blockblame/blockblame.go:26]: let's use `*Manufacturer` here since that's what you return below

File: net/tlsdial/tlsdial.go [hunk 2]
```diff
@@ -92,6 +103,12 @@ func Config(host string, ht *health.Tracker, base *tls.Config) *tls.Config {
 		}
 		if ht != nil {
 			defer func() {
+				if m, found := blockblame.VerifyCertificate(cs.PeerCertificates[0]); found {
```
Comment (bradfitz) [inline net/tlsdial/tlsdial.go:106]: s/found/ok/

File: net/tlsdial/blockblame/blockblame.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import (
+	"crypto/x509"
+	"strings"
+)
+
+// Manufacturer represents a firewall manufacturer that may be blocking Tailscale.
+type Manufacturer struct {
+	// Name of the firewall manufacturer, e.g. "Fortinet"
+	CompanyName string
+	// List of indicators that can be used to identify the manufacturer
```
Comment (bradfitz) [inline net/tlsdial/blockblame/blockblame.go:15]: // Indicators is ...

But how about instead:

    // Match reports whether a cert matches this Manufacturer.
   Match func(*x509.Certificate) bool

Then you can have helper functions `func any` and `hasOUSubstr` etc.

File: net/tlsdial/blockblame/blockblame.go (hunk 1 above)
Comment (agottardo) [reply net/tlsdial/blockblame/blockblame.go:26]: Done.

File: net/tlsdial/blockblame/blockblame.go (hunk 3 above)
Comment (agottardo) [reply net/tlsdial/blockblame/blockblame.go]: done

File: net/tlsdial/tlsdial.go (hunk 2 above)
Comment (agottardo) [reply net/tlsdial/tlsdial.go]: done

File: net/tlsdial/blockblame/blockblame.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+		if m.match != nil && m.match(c) {
+			return m, true
+		}
+	}
+	return nil, false
+}
+
+// Manufacturer represents a firewall manufacturer that may be blocking Tailscale.
+type Manufacturer struct {
+	// Name of the firewall manufacturer, e.g. "Fortinet"
```
Comment (bradfitz) [inline net/tlsdial/blockblame/blockblame.go:28]: // Name is the name of the firewall manufacturer to be
// mentioned in health warning messages.

(complete sentence: https://go.dev/wiki/CodeReviewComments#comment-sentences)

File: net/tlsdial/blockblame/blockblame.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	CompanyName string
```
Comment (bradfitz) [inline net/tlsdial/blockblame/blockblame.go:29]: Name string

(CompanyName is unnecessarily long)

File: net/tlsdial/blockblame/manufacturers.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package blockblame
```
Comment (bradfitz) [inline net/tlsdial/blockblame/manufacturers.go:4]: I'd just merge this file in with the other one, really.

File: net/tlsdial/tlsdial.go [hunk 7]
```diff
@@ -44,6 +45,16 @@ var debug = envknob.RegisterBool("TS_DEBUG_TLS_DIAL")
 // Headscale, etc.
 var tlsdialWarningPrinted sync.Map // map[string]bool
 
+var mitmBlockWarnable = *health.Register(&health.Warnable{
```
Comment (bradfitz) [inline net/tlsdial/tlsdial.go:48]: drop the `*` deference here. We don't do that elsewhere.

Comment (agottardo) [reply net/tlsdial/tlsdial.go:48]: done

File: net/tlsdial/blockblame/manufacturers.go (hunk 6 above)
Comment (agottardo) [reply net/tlsdial/blockblame/manufacturers.go:4]: done

File: net/tlsdial/blockblame/blockblame.go (hunk 5 above)
Comment (agottardo) [reply net/tlsdial/blockblame/blockblame.go:29]: done

File: net/tlsdial/tlsdial.go [hunk 8]
```diff
@@ -86,12 +97,29 @@ func Config(host string, ht *health.Tracker, base *tls.Config) *tls.Config {
...
+			}
 		}
 		if ht != nil {
 			defer func() {
+				if retErr != nil && cert != nil {
+					// Is it a MITM SSL certificate from a well-known network appliance manufacturer?
+					// Show a dedicated warning.
+					m, ok := blockblame.VerifyCertificate(cert)
+					if ok {
+						log.Printf("tlsdial: server cert for %q looks like %q equipment (blocking Tailscale!)", host, m.Name)
```
Comment (bradfitz) [inline net/tlsdial/tlsdial.go:115]: drop the exclamation mark. they don't age well.

(everybody thinks THEIR! log messages need exclamation marks, but then they all do. better that none do.)

File: net/tlsdial/blockblame/manufacturers.go [hunk 9]
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package blockblame
```
Comment (bradfitz) [inline net/tlsdial/blockblame/manufacturers.go:4]: empty file; delete

**Review by bradfitz — APPROVED** [review]

**Pull request merged**